syntax = "proto3";

package auth.v1;

option go_package = "erp.localhost/internal/infra/proto/generated/auth/v1;authv1";

import "google/protobuf/timestamp.proto";
import "infra/v1/infra.proto";

// ============================================================================
// Resource Definitions (Role, Permission, etc.)
// ============================================================================

message RoleMetadata {
    string color = 1;
    string icon = 2;
    int32 max_assignable = 3;
}

message RoleData {
    string id = 1;
    string tenant_id = 2;
    string name = 3;
    string slug = 4;
    string description = 5;
    string type = 6;                                // "system" or "custom"
    bool is_tenant_admin = 7;                       // Tenant admin role flag
    repeated string permissions = 8;
    int32 priority = 9;
    string status = 10;                             // "active" or "inactive"
    RoleMetadata metadata = 11;
    google.protobuf.Timestamp created_at = 12;
    google.protobuf.Timestamp updated_at = 13;
    string created_by = 14;
}

message PermissionData {
    string id = 1;
    string tenant_id = 2;
    string name = 3;
    string slug = 4;
    string description = 5;
    string resource = 6;                            // e.g., "users", "orders"
    string action = 7;                              // e.g., "create", "read", "update", "delete"
    string status = 8;
    google.protobuf.Timestamp created_at = 9;
    google.protobuf.Timestamp updated_at = 10;
    string created_by = 11;
}

// ============================================================================
// Create Resource Request
// ============================================================================

message CreateRoleData {
    string tenant_id = 1;
    string name = 2;
    string slug = 3;
    string description = 4;
    string type = 5;
    bool is_tenant_admin = 6;                       // Tenant admin role flag
    repeated string permissions = 7;
    int32 priority = 8;
    string status = 9;
    RoleMetadata metadata = 10;
    string created_by = 11;
}

message CreatePermissionData {
    string tenant_id = 1;
    string name = 2;
    string slug = 3;
    string description = 4;
    string resource = 5;
    string action = 6;
    string status = 7;
    string created_by = 8;
}

// ============================================================================
// Update Resource Request
// ============================================================================

message UpdateRoleData {
    string id = 1;
    string tenant_id = 2;
    optional string name = 3;
    optional string slug = 4;
    optional string description = 5;
    optional string type = 6;
    optional bool is_tenant_admin = 7;              // Tenant admin role flag
    repeated string permissions = 8;
    optional int32 priority = 9;
    optional string status = 10;
    optional RoleMetadata metadata = 11;
}

message UpdatePermissionData {
    string id = 1;
    string tenant_id = 2;
    optional string name = 3;
    optional string slug = 4;
    optional string description = 5;
    optional string resource = 6;
    optional string action = 7;
    optional string status = 8;
}


// ============================================================================
// User Role Assignment
// ============================================================================

message AssignRolesRequest {
    infra.v1.UserIdentifier identifier = 1;        // User to assign roles to
    repeated string role_ids = 2;                  // Role IDs to assign
    string assigned_by = 3;                        // Who is assigning the roles
}

message RemoveRolesRequest {
    infra.v1.UserIdentifier identifier = 1;        // User to remove roles from
    repeated string role_ids = 2;                  // Role IDs to remove
    string removed_by = 3;                         // Who is removing the roles
}


// ============================================================================
// Dedicated Service Request/Response Messages
// ============================================================================

// Role Service Messages
message CreateRoleRequest {
    infra.v1.UserIdentifier identifier = 1;        // Requestor identity
    CreateRoleData role = 2;                       // Role data to create
    string target_tenant_id = 3;                   // Target tenant (for cross-tenant operations)
}

message CreateRoleResponse {
    string role_id = 1;
}

message UpdateRoleRequest {
    infra.v1.UserIdentifier identifier = 1;        // Requestor identity
    UpdateRoleData role = 2;                       // Role data to update
    string target_tenant_id = 3;                   // Target tenant (for cross-tenant operations)
}

message GetRoleRequest {
    infra.v1.UserIdentifier identifier = 1;        // Requestor identity
    string role_id = 2;                            // Role ID to retrieve
    string target_tenant_id = 3;                   // Target tenant (for cross-tenant operations)
}

message GetRoleResponse {
    RoleData role = 1;
}

message ListRolesRequest {
    infra.v1.UserIdentifier identifier = 1;        // Requestor identity
    string target_tenant_id = 2;                   // Target tenant (for cross-tenant operations)
    optional infra.v1.PaginationRequest pagination = 3;
}

message ListRolesResponse {
    repeated RoleData roles = 1;
    infra.v1.PaginationResponse pagination = 2;
}

message DeleteRoleRequest {
    infra.v1.UserIdentifier identifier = 1;        // Requestor identity
    string role_id = 2;                            // Role ID to delete
    string target_tenant_id = 3;                   // Target tenant (for cross-tenant operations)
}

// Permission Service Messages
message CreatePermissionRequest {
    infra.v1.UserIdentifier identifier = 1;        // Requestor identity
    CreatePermissionData permission = 2;           // Permission data to create
    string target_tenant_id = 3;                   // Target tenant (for cross-tenant operations)
}

message CreatePermissionResponse {
    string permission_id = 1;
}

message UpdatePermissionRequest {
    infra.v1.UserIdentifier identifier = 1;        // Requestor identity
    UpdatePermissionData permission = 2;           // Permission data to update
    string target_tenant_id = 3;                   // Target tenant (for cross-tenant operations)
}

message GetPermissionRequest {
    infra.v1.UserIdentifier identifier = 1;        // Requestor identity
    string permission_id = 2;                      // Permission ID to retrieve
    string target_tenant_id = 3;                   // Target tenant (for cross-tenant operations)
}

message GetPermissionResponse {
    PermissionData permission = 1;
}

message ListPermissionsRequest {
    infra.v1.UserIdentifier identifier = 1;        // Requestor identity
    string target_tenant_id = 2;                   // Target tenant (for cross-tenant operations)
    optional infra.v1.PaginationRequest pagination = 3;
}

message ListPermissionsResponse {
    repeated PermissionData permissions = 1;
    infra.v1.PaginationResponse pagination = 2;
}

message DeletePermissionRequest {
    infra.v1.UserIdentifier identifier = 1;        // Requestor identity
    string permission_id = 2;                      // Permission ID to delete
    string target_tenant_id = 3;                   // Target tenant (for cross-tenant operations)
}

// Verification Service Messages
message CheckPermissionsRequest {
    infra.v1.UserIdentifier identifier = 1;        // User to check
    repeated string permissions = 2;               // Permissions to check
}

message CheckPermissionsResponse {
    map<string, bool> permissions = 1;             // Map of permission -> has_permission
}

message HasPermissionRequest {
    infra.v1.UserIdentifier identifier = 1;        // User to check
    string permission = 2;                         // Permission to check
    string target_tenant_id = 3;                   // Target tenant (for cross-tenant operations)
}

message HasPermissionResponse {
    bool has_permission = 1;
}

message GetUserPermissionsRequest {
    infra.v1.UserIdentifier identifier = 1;        // User to check
}

message GetUserPermissionsResponse {
    map<string, bool> permissions = 1;             // Map of all user permissions
}

message GetUserRolesRequest {
    infra.v1.UserIdentifier identifier = 1;        // User to check
}

message GetUserRolesResponse {
    repeated string role_ids = 1;                  // List of role IDs
}

message IsSystemTenantUserRequest {
    string tenant_id = 1;                          // Tenant ID to check
}

message IsSystemTenantUserResponse {
    bool is_system_tenant = 1;
}

// ============================================================================
// Dedicated Service Definitions
// ============================================================================

// RoleService provides role management operations
service RoleService {
    rpc CreateRole(CreateRoleRequest) returns (CreateRoleResponse);
    rpc UpdateRole(UpdateRoleRequest) returns (infra.v1.Response);
    rpc GetRole(GetRoleRequest) returns (GetRoleResponse);
    rpc ListRoles(ListRolesRequest) returns (ListRolesResponse);
    rpc DeleteRole(DeleteRoleRequest) returns (infra.v1.Response);
}

// PermissionService provides permission management operations
service PermissionService {
    rpc CreatePermission(CreatePermissionRequest) returns (CreatePermissionResponse);
    rpc UpdatePermission(UpdatePermissionRequest) returns (infra.v1.Response);
    rpc GetPermission(GetPermissionRequest) returns (GetPermissionResponse);
    rpc ListPermissions(ListPermissionsRequest) returns (ListPermissionsResponse);
    rpc DeletePermission(DeletePermissionRequest) returns (infra.v1.Response);
}

// VerificationService provides permission and role verification operations
service VerificationService {
    rpc CheckPermissions(CheckPermissionsRequest) returns (CheckPermissionsResponse);
    rpc HasPermission(HasPermissionRequest) returns (HasPermissionResponse);
    rpc GetUserPermissions(GetUserPermissionsRequest) returns (GetUserPermissionsResponse);
    rpc GetUserRoles(GetUserRolesRequest) returns (GetUserRolesResponse);
    rpc IsSystemTenantUser(IsSystemTenantUserRequest) returns (IsSystemTenantUserResponse);
}